### DISCUSSION TITLE
Weighted bodies, following joints in gLTF?

### DISCUSSION THREAD

**Discussion by ThomasCartier** [jrouwe/JoltPhysics#804] [Q&A] [answered]
Hi,

I succeeded to make everything rolling using the ragdoll. Now, I would like to go one step further.
Do we have a way to modulate the shape to remain as realistic as possible, using [a joint matrix](https://github.com/KhronosGroup/glTF-Tutorials/blob/master/gltfTutorial/gltfTutorial_020_Skins.md#the-joint-matrices) like we would in gltf to have the Shape as close/realistic as possible from the original mesh ?

It implies to modify the attached Shape step by step to have vertices perfectly in place.

Right now, once the shape is set up and attached to a ragdoll part, it cannot be changed at the vertex level to apply the differing weights (maybe I am wrong). The position/rotation applies to the whole object using `BodyInterface::MoveKinematic ` and `MotionProperties::MoveKinematic`.

Strategies I can see here:

1. Making `BodyInterface::MoveKinematic` work at vertex level using a new weight field to apply the new position to each vertex. But I am not sure if we would not run into a tunneling effect if a collider was already so close and the shape kind of expands (think a cloth "point" or a rigid stuck in the newly modified shape). Maybe lot of code change.
2. If, like me, the ragdoll is only kinematic: precalculate all the parts of the shapes and activate them at the right step. That may be a lot of parts to manage, but it is what it is (and not hard). That would guarantee to not touch the current code, ditch the ragdoll and skeleton altogether, and have a perfect matching body collision for the animation. But I fear about the same issue as above with trapped collisions when two bodies where so close and a completely new shape that activates: in case of cloth, I may not now that it should be pushed away if the new shape inflates even just a bit. I would like to make sure we can avoid the tunneling effect. 
3. Same as above, but instead of activating a different shape, tell the physic system what the next shape for a given shape will be so that I can make sure that the simulation looks fine.

Maybe there is something I missed, but I am targeting a realism as high as I can in this game sequence; Pre-baking is an option ( I could make it a quick time event sequence for a best looking result).

Is it guaranteed that at least options {2-3} are a viable approach ? Did I miss a simpler approach that's already there ?

**Comment by jrouwe** [comment]
What you're describing is not a rigid body simulation. In a rigid body simulation, the body itself remains constant in shape. You can switch to a soft body and animate the velocities of the vertices to reach the desired skinned position in the next physics step: each vertex gets velocity = (next_skinned_position - current_simulated_position) / time_step_delta_time every frame. That would ensure it can push dynamic objects away. A simple example is here: SoftBodyKinematicTest.

I have to say, to do it this way would be pretty unconventional and CPU intensive. Usually we accept that there is some discrepancy between the skinned mesh and the physics sim so we model body parts as simple (tapered) capsules attached to joints.

**Reply by ThomasCartier** [reply]
Nice thank you, but I read that soft soft collision is not yet part of the code base, right? (Again, thanks for your valuable info)

**Reply by jrouwe** [reply, marked as answer]
No soft vs soft is not implemented, but if your 'ragdoll' is keyframed then it won't collide with another keyframed ragdoll anyway, it will only collide with dynamic bodies (which keyframed soft bodies will also collide with).

**Reply by ThomasCartier** [reply]
Perfect! Thanks!

**Reply by ThomasCartier** [reply]
Hi!

So I could try multiple ways to do it, and the kinematic works by itself, but what I wanted is the following:
![sbs](https://github.com/jrouwe/JoltPhysics/assets/146366386/63ba120e-a43a-448c-9a3d-83f88077e08f)

(This is a simpler scene with simpler shapes for the question).

The cube and the sphere are both soft bodies. Animating the cube and the sphere via kinematics independently works just fine. But their collision is not taken into account, even if the cube has its vertex velocities updated at each frame.

the sphere is just soft body while the cube is a keyframed soft body.

Retrying with a rigid cube and the collision works. So "keyframed soft" vs soft doesn't work here.
Is that expected? Do we land in the territory of Soft body vs soft body collision ?

**Reply by jrouwe** [reply]
A keyframed soft body is still a soft body, so it cannot collide with other soft bodies.

**Reply by ThomasCartier** [reply]
Indeed, thanks Jorrit, 👍
